### PR TITLE
[MM-44159] Members RHS: fix search + allow to load more users

### PR DESCRIPTION
**Pull request by JulienTant**
#### Summary
This PR fixes the search two issues:

- Be able to load more than 100 users at a time using infinite scrolling + a backup button.
- Searching was not working on not loaded users.

```diff
- IMPORTANT NOTICE:
```

This is a 'poor-man' solution as it does not take into consideration the performance issue a user would face if they decide to load thousands of users in the list. It is difficult to use a virtualized list to solve the issue because the scrollable zone is composed of two lists:

image

As a solution we could:
- Add simple pagination instead of infinite scrolling on the members list.
- Making only the "Members" area scrollable, but we could end up with weird-looking pages if the channel admin section is pretty height.

In both scenarios, it will also require being able to load 'Channel admin' and 'Members' separately instead of all together followed by a sorting process.

If there's a better solution I'm happy to hear it.

I decided to submit this PR anyway to propose a 'quick solution' to the problem, if we would rather not offer this solution I'm fine closing this PR and delaying it until we have a better solution - and more time on my side to work on it.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-44159

#### Related Pull Requests
- Has server changes (please link here)
- Has mobile changes (please link here)

#### Screenshots

https://user-images.githubusercontent.com/785518/169131868-d933c44d-53ba-4cc9-9847-ff710a4ad908.mp4

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by matthewbirtch** [comment]
Is there a way to load up a lot of test users on this server so we can see it in action @JulienTant ?

**Comment by JulienTant** [comment]
> Is there a way to load up a lot of test users on this server so we can see it in action @JulienTant ?

Yep I loaded 1000 users!

**Review by matthewbirtch — APPROVED** [review]
Seems pretty great to me! Not noticing any performance issues

**Comment by JulienTant** [comment]
> Seems pretty great to me! Not noticing any performance issues

It's going to be computer-dependent and the more you load the worst it's getting.

File: components/channel_members_rhs/channel_members_rhs.tsx [hunk 1]
```diff
@@ -88,10 +106,32 @@ export default function ChannelMembersRHS({channel, searchTerms, membersCount, c
...
         setEditing(false);
-        actions.loadProfilesAndReloadChannelMembers(channel.id);
+        setPage(0);
+        setLoadingMore(false);
+
+        actions.loadProfilesAndReloadChannelMembers(0, USERS_PER_PAGE, channel.id);
         actions.loadMyChannelMemberAndRole(channel.id);
     }, [channel.id, channel.type]);
 
+    const searchDebounced = debounce(
```
Comment (hmhealey) [inline components/channel_members_rhs/channel_members_rhs.tsx:116]: This is going to get redefined every time this re-renders, so you'll lose the state tracking the debouncing. I'm not sure exactly how to make this work in a hook, but maybe something like wrapping it in a `useMemo` would work to fix that?

File: components/channel_members_rhs/channel_members_rhs.tsx [hunk 2]
```diff
@@ -88,10 +106,32 @@ export default function ChannelMembersRHS({channel, searchTerms, membersCount, c
...
+        async () => {
+            if (searchTerms) {
+                setEditing(false);
+                setPage(0);
+                setLoadingMore(false);
+
+                await actions.searchProfilesAndChannelMembers(searchTerms, {in_team_id: channel.team_id, in_channel_id: channel.id});
+                actions.loadProfilesAndReloadChannelMembers(0, USERS_PER_PAGE, channel.id);
+                actions.loadMyChannelMemberAndRole(channel.id);
```
Comment (hmhealey) [inline components/channel_members_rhs/channel_members_rhs.tsx:125]: Do these need to be reloaded when we're searching? I would think we only need the search results at this point

File: components/channel_members_rhs/channel_members_rhs.tsx [hunk 3]
```diff
@@ -122,6 +162,28 @@ export default function ChannelMembersRHS({channel, searchTerms, membersCount, c
         await actions.closeRightHandSide();
     };
 
+    const loadMore = async () => {
+        const nextPage = page + 1;
+        setPage(nextPage);
+        setLoadingMore(true);
+
+        await actions.loadProfilesAndReloadChannelMembers(nextPage, USERS_PER_PAGE, channel.id);
+        await actions.loadMyChannelMemberAndRole(channel.id);
```
Comment (hmhealey) [inline components/channel_members_rhs/channel_members_rhs.tsx:171]: Similarly, why do we need to reload the current user's channel member each time?

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: i18n/en.json [hunk 4]
```diff
@@ -2808,6 +2808,8 @@
   "channel_members_rhs.header.title": "Members",
   "channel_members_rhs.list.channel_admin_title": "CHANNEL ADMINS",
   "channel_members_rhs.list.channel_members_title": "MEMBERS",
+  "channel_members_rhs.list.load-more": "Load More",
+  "channel_members_rhs.list.loading": "Loading",
```
Comment (julmondragon) [inline i18n/en.json:2812]: nit: Could add the ellipsis char `…` or use the `LoadingScreen` component.

**Review by julmondragon — CHANGES_REQUESTED** [review]
Noticed some behaviors:

1. After reaching the end of the list, the scroll continues firing requests for more data not available:

https://user-images.githubusercontent.com/79058848/169321797-9ef2026a-7f50-412b-a579-1f0c3ae789d8.mp4

2. After reaching the end of the list, going back to channel info and then back again to the members list, the full list is already loaded, but the scroll fires requests to load data since the beginning (page 1):

https://user-images.githubusercontent.com/79058848/169321902-6ab79311-18b7-46e0-88ae-30c84ad22b9d.mp4

3. After having loaded a big amount of users the actions start to lag a little bit.

**Comment by JulienTant** [comment]
Staling this PR for now; I don't have enough bandwidth

**Pull request closed without merging**